untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,549 @@
+    That is the result of the computation.
+
+<hr>
+
+Any expression where beta-reduction can be applied, we'll call a redex (short for reducible expression).
+
+For example, the expression $(\lambda x. x) y$ is a redex, and one step of beta-reduction transforms it to just $y$.
+
+According to the rule, we rename bound variables if there might be ambiguity.
+For example, $(\lambda x. \lambda y. x\;y)\;(\lambda x. x\;y)$ is a redex.
+However, the $y$ identifier is ambiguous.
+In the left term, it refers to the bound variable; in the right one, it refers to a free variable.
+
+To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$:
+
+$(\lambda x. \lambda z. x\;z)\;(\lambda x. x\;y).$
+
+Then we can do the rest of the steps with no fear of confusion:
+
+$\lambda z. (\lambda x. x\;y)\;z.$
+
+In the last example, you might wonder if we should also apply beta-reduction inside the abstraction body.
+This brings us to the discussion of evaluation strategies.
+
+### Evaluation strategies
+
+An evaluation strategy is a rule that defines which redexes are reduced and in what order.
+
+The simplest approach is to reduce redexes in any random order until there's nothing left.
+This approach is called _full beta-reduction._
+But it has some issues.
+For instance, you might get different results depending on the reduction order.
+
+So it's better to reduce in a specific order.
+There are multiple ways to approach this. 
+
+Our decision points are: 
+* to start with the left or right side of the expression;
+* to start with the outermost or the innermost redex;
+* to reduce redexes inside abstraction bodies or not.
+
+#### Direction of reduction
+
+We can reduce an expression left-to-right or right-to-left.
+This decision is easy: going left-to-right is strictly more powerful because the process terminates on all terms for which going right-to-left does and then some. 
+
+#### Reducing outside-in or inside-out
+
+Second, we can choose to first reduce the outermost or the innermost redex.
+The outermost redex is a redex not contained within any other redex, and the innermost redex is a redex that does not contain any other redexes.
+
+To see what we mean, consider the following expression:
+
+$(\lambda x. x)\;((\lambda y. y)\;z).$
+
+This expression contains two redexes, itself and $(\lambda y. y)\;z.$
+The former is not contained within any other redex, so it is the outermost of the two.
+The sub-expression does not contain any other redex, so is the innermost.
+
+Reducing the leftmost outermost redex first is called the _normal order strategy_, and reducing the leftmost innermost first is called the _applicative order strategy_.
+
+In other words, the applicative order strategy first reduces all of the arguments (from left to right) and then applies the abstraction, while normal order does the inverse.
+
+Expressions that are not reducible using the chosen evaluation strategy are said to be in the _normal form_.
```

Comment:
@lierdakil Since it mentions only strong strategies, I decided to put the comparison here. If it is more of a general statement on the superiority of outermost-first reducing, perhaps it should be at the end, but then it should also somehow mention the weak strategies, I think.